MozillaFoundation/foundation.mozilla.org
# Custom base for `LinkBlock`

# Description



#11954 changed `LinkBlock` to handle multiple types of links by introducing `wagtail-link-block` as a dependency to manage multiple link types. `wagtail-link-block` seemed like a good solution for standardising the way we handle links, since it was ready-made and was a good fit to our needs.

However, as I was defining a `NavLinkBlock` (#12006) that only allowed for pages, external urls and emails to use in our navigation menu, I ran into several limitations with `wagtail-link-block`. Creating a child of `wagtail-link-block` while removing fields somehow seemed like an anti-pattern and in any case its code did not handle that well.

The level of customisation necessary to remove the extra fields would make it almost a complete rewrite.

Hence, to have a more scalable approach that works for both the current `LinkBlock` and the future `NavLinkBlock`s (used in the main nav dropdowns), I decided to recreate `wagtail-link-block` supporting a dynamic number of fields.

## Implementation

This PR introduces a `BaseLinkBlock` with links to:
- Page
- External URL
- Email
- Anchor

The `BaseLinkBlock` is an abstract class, so it won't be recognised by Django/Wagtail on its own.

The only concrete implementation right now is on `LinkBlock`, which extends `BaseLinkBlock` to include also file and phone links.

This same `BaseLinkBlock` can be reused later for the `NavLinkBlock`.

## Production data

This work has also modified the migrations in #11954 (not deployed yet) to work with this new approach. The migrations will convert all current links in production to an "external URL" type of link.

## Testing

1. Log in to the CMS
2. Go to "Reports -> Block types" and find the `LinkBlock` column
3. Select a page type on in the `LinkBlock` column and verify that their links have been migrated to the new interface
4. Verify that you can add a new link block to the relevant field for that page type
5. Repeat steps 3 and 4 for all page types on the `LinkBlock` column


Link to sample test page:
Related PRs/issues: 
- #12006 
- #11954
- https://mozilla-hub.atlassian.net/browse/TP-292

# Checklist



**Tests**

**Changes in Models:**
- ~~[ ] Are my changes [backward-compatible](https://github.com/mozilla/foundation.mozilla.org/blob/main/docs/workflow.md#django-migrations-what-to-do-when-working-on-backward-incompatible-migrations). If not, did I schedule a deploy with the rest of the team?~~ :warning: This will result in downtime. preboot should be **disabled** :warning:

**Documentation:**

**Merge Method**
**💡❗Remember to use squash merge when merging non-feature branches into `main`**

┆Issue is synchronized with this [Jira Story](https://mozilla-hub.atlassian.net/browse/TP-316)
